[Onboarding] Search Artists / Categories

mennenia opened the pull request:
TODO:
- [x] iPad layout optimisation (currently working on)
- [x] Categories/Gene API (see #1454)
- [x] Storing of intermediate follow IDs (pending decision on method)
- [x] `weak` hold to network requests

Required:
@katarinabatina 
- Could I get the search icon assets in a medium gray please? 

Demo party:

![demo](https://cloud.githubusercontent.com/assets/373860/14779078/43772e42-0ace-11e6-8173-0515316ee7ff.GIF)

ArtsyMobile commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">1 Error</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:no_entry_sign:</td>
      <td>No CHANGELOG changes made</td>
    </tr>
    
  </tbody>
</table>
  

  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">2 Warnings</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td>PR is classed as Work in Progress</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td>Needs testing on a Phone if change is non-trivial</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

alloy commented:
Looking awesome!

Regarding the animation that replaces the followed artist, I find it to be too subtle personally. Is it easy (like a few minutes work) to replace the cross-fade animation with something wild like a left-out & right-in animation?

alloy commented on Artsy/Models/API_Models/Artwork_Metadata/Gene.m:
```diff
@@ -79,6 +79,11 @@ - (BOOL)isFollowed
+- (NSString *)id
+{
+    return _geneID;
+}
```
Might be better to have these models implement something like the `ARShareableObject` protocol? (Even though this is not actually about sharing.) They would only need to implement the `publicArtsyID` method, though.

Also, only use the ivars directly from `initXXX`, `dealloc`, or the property’s setter/getter methods.

alloy replied:
Oh wait, I now see you chose the `ARFollowable` protocol, that’s indeed a much better choice. Just change the access to use the getter methods instead of direct ivar access.

alloy commented on Artsy/Protocols/ARFollowable.h:
```diff
@@ -3,6 +3,7 @@
 @protocol ARFollowable <NSObject>
 
 @property (nonatomic, assign, getter=isFollowed) BOOL followed;
+@property (nonatomic, strong, readonly) NSString *id;
```
For consistency reasons, it might still be a good idea to use the same getter as `ARShareable` uses. i.e. `publicArtsyID`.

mennenia replied:
Noted! Am considering taking this out though. If we're storing the full objects in a set, I can just call the `follow` methods on them without needing the ID separately.

alloy replied:
Hmm, this leads to me wonder if there’s an existing way to follow a batch of artists/categories instead of having to make many requests for each individual entity. Can you add it to your API requirements ticket, if you think that makes sense?

mennenia replied:
Will investigate and update if doesn't exist. 

Also, probably will need the ID later on to send the requests excluding what we're showing, etc but I'll deal with it when that happens. In that case, I'll make sure to follow the `ARShareable` pattern. Removed for now.

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/AROnboardingViewController.h:
```diff
@@ -3,12 +3,14 @@
 - (void)personalizeCategoriesDone;
 - (void)personalizeBudgetDone;
 - (void)backTapped;
+- (void)followableItemFollowed:(NSObject<ARFollowable> *)item;
```
The cocoa convention is to use `id<ARFollowable>`. Usually the protocol will pull in the `NSObject` protocol, as does `ARFollowable`.

mennenia replied:
Doh, yup. Fixin'!

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/Onboarding_stages/5-_Onboarding_Personalization/AROnboardingFollowableTableViewCell.m:
```diff
@@ -0,0 +1,65 @@
+
+- (void)updateConstraints
+{
+    [super updateConstraints];
```
Is this left in by accident?

mennenia replied:
Yup - removing

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/Onboarding_stages/5-_Onboarding_Personalization/AROnboardingPersonalizeTableViewController.m:
```diff
@@ -0,0 +1,202 @@
+        } completion:^(BOOL finished){
+        }];
+    }
+}
```
Intens! This is for the initial staggered loading of the list, right? Could you add some comments explaining what the individual lines do and what the magic numbers are based on?

mennenia replied:
It's for the "You might also like" staggered animation. Will definitely comment. Magic is mostly just magic though.

alloy replied:
Fair 👍

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/Onboarding_stages/5-_Onboarding_Personalization/AROnboardingPersonalizeTableViewController.m:
```diff
@@ -0,0 +1,202 @@
+    if ([result isKindOfClass:[Artist class]]) {
+        Artist *artist = (Artist *)result;
+        cell.title.text = artist.name;
+        cell.thumbnail.image = [UIImage imageWithData:[NSData dataWithContentsOfURL:artist.squareImageURL]];
```
You should use these category methods on `UIImageView` which use SDWebImage to download and cache images https://github.com/artsy/eigen/blob/master/Artsy/Categories/Apple/UIImageView+AsyncImageLoading.h

mennenia replied:
👍 Thanks!

Might explain why the list is sometimes a little bit slow.

alloy replied:
Yeah totally, the `NSData` route you chose is synchronous.

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/Onboarding_stages/5-_Onboarding_Personalization/ARPersonalizeViewController.m:
```diff
@@ -230,273 +85,209 @@ - (void)showViews
             break;
         case AROnboardingStagePersonalizeBudget:
-            [self.headerView setupHeaderViewWithTitle:@"Do you have a budget in mind?" andProgress:70.0f];
+            [self.headerView setupHeaderViewWithTitle:@"Do you have a budget in mind?" andProgress:0.7f];
```
@mennenia @sarahscott @orta @ashfurrow We should probably get rid of the `f` suffix for `CGFloat` types, because it’s meant to tell the compiler to use single precision, i.e. `float`, whereas `CGFloat` on 64-bit machines has double precision, i.e. `double`.

mennenia replied:
I didn't even know that was what it was for. Thanks! And will remove, makes sense.

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/Onboarding_stages/5-_Onboarding_Personalization/ARPersonalizeViewController.m:
```diff
@@ -230,273 +85,209 @@ - (void)showViews
-        return;
+    if (self.searchRequestOperation) {
+        [self.searchRequestOperation cancel];
     }
```
Another trick I started using is to rely on AFNetworking to retain the operation as long as it is in-flight, which means that you can make your property `weak` and thus you can just send it the `cancel` message without checking if it still exists.

No need to tackle that right now, it’s just an FYI.

alloy replied:
Additionally, using `weak` means it’s also released as soon as possible, right now I haven’t seen the code (yet) where you release the operation.

mennenia replied:
Ohh, that's good to know! I will definitely switch to that. Will be honest, I followed what seemed to be conventional in the class before, and I may have missed the release part.

alloy replied:
It was probably just being released as part of the VCs dealloc and isn’t a big deal as it’s just a smallish object, but hey if you can prevent it altogether ¯\_(ツ)_/¯

I think @sarahscott ended up using this `weak` pattern in the WorksForYou network model.

mennenia commented:
Addressed comments! Will check about the API, and fix the `weak` thing. 

As for the fade animation, it's what was requested in the design spec. I'll bring it up when I do QA with @katarinabatina later this week! :)

alloy commented:
> As for the fade animation, it's what was requested in the design spec. I'll bring it up when I do QA with @katarinabatina later this week! :)

Sure, I get that, I’d just like to see what other options look like, which also makes it easier for you to discuss it during QA.

alloy commented on Artsy/View_Controllers/Login_and_Onboarding/Onboarding_stages/5-_Onboarding_Personalization/AROnboardingPersonalizeTableViewController.m:
```diff
@@ -132,13 +133,22 @@ - (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger
+
+        // The lower down the cell is in the list, the longer the animation takes (cell has further to travel)
+        // Equally, to stagger, we want the bottom cells to start their animation later than the top cells
+        // Other than that, magic constants are just the result of tweaks over time
```
Great comments 👍 :100:

mennenia commented:
Preview on the iPad layouts:

<img width="1023" alt="screen shot 2016-04-25 at 17 04 56" src="https://cloud.githubusercontent.com/assets/373860/14790130/d4785e0a-0b07-11e6-9984-6decf881c494.png">

Known issues: 
- progress bar should be full width
- border missing on "next" and "back buttons"
- search icon is white instead of grey

These will be address in sequence polish, as it has more to do with that than search / the tableview. 

As far as I'm concerned I'm happy with this. Let me know if there are any further comments / issues!

alloy commented:
Looks good, code wise 👍
